torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -386,9 +406,26 @@ void RRefContext::delPendingUser(const ForkId& forkId) {
   TORCH_INTERNAL_ASSERT(
       iter != pendingUsers_.end(),
       "Inconsistent states: attempt to delete a non-exist UserRRef.");
+  // unblock pending user functions
+  iter->second->confirm();
   pendingUsers_.erase(iter);
 }
 
+void RRefContext::recordThreadLocalPendingUsers() {
+  TORCH_INTERNAL_ASSERT(
+      userTable_.empty(),
+      "User RRef Table should be empty when start recording");
+  recording = true;
+}
+
+void RRefContext::waitForThreadLocalPendingUsers() {
+  for (auto& state: userTable_) {
+    state->future_.wait();
+  }
+  userTable_.clear();
+  recording = false;
```

Comment:
Why do we want to stop recording additional pending users once we've called `waitForThreadLocalPendingUsers`?

Reply:
pending users are added into the context for two reasons:

1. when we call `rpc.remote`, the return value `UserRRef` on the caller will be added to the context.
2. when we forward an RRef to another user, the `UserRRef` on the destination will be added to the context. 

For 1, we don't block as we would like to use that `UserRRef` right away. We only block in 2. So, if we don't close the recording session, there could be more `UserRRef`s added into the `userTable_`. Consider a case where the application runs a user function remotely, and the user function creates a RRef using `rpc.remote`. This RRef should not block the next user function handled by the same RPC thread. 

Thanks for pointing this out, let me add this description to the code.